dex/src/main/java/io/crate/data/NestedLoopBatchIterator.java

```diff
@@ -0,0 +1,506 @@
+            this.joinCondition = joinCondition.apply(rowData());
+        }
+
+        @Override
+        public boolean moveNext() {
+            if (loading) {
+                throw new IllegalStateException("BatchIterator is loading");
+            }
+            while (true) {
+                pointToCombinedColumns();
+                if (activeIt == left) {
+                    return moveLeftSide();
+                }
+
+                Boolean x = tryAdvanceRight();
+                if (x != null) {
+                    return x;
+                }
+                activeIt = left;
+            }
+        }
+
+        private boolean moveLeftSide() {
+            activeIt = right;
+            while (left.moveNext()) {
+                Boolean x = tryAdvanceRight();
+                if (x != null) {
+                    return x;
+                }
+            }
+            activeIt = left;
+            return false;
+        }
+
+        @Override
+        public CompletionStage<?> loadNextBatch() {
+            loading = true;
+            return super.loadNextBatch().whenComplete((result, failure) -> loading = false);
+        }
+
+        /**
+         * try to move the right side
+         * @return true  -> moved and matched
+         *         false -> need to load more data
+         *         null  -> reached it's end and moved back to start -> left side needs to continue
+         */
+        private Boolean tryAdvanceRight() {
+            while (right.moveNext()) {
+                if (joinCondition.getAsBoolean()) {
+                    hadMatch = true;
+                    return true;
+                }
+            }
+            if (right.allLoaded() == false) {
+                return false;
+            }
+            right.moveToStart();
+            if (hadMatch == false) {
+                activeIt = left;
+                pointToNullColumns();
+                return true;
+            }
+            hadMatch = false;
+            return null;
+        }
+
+        // TODO: inputs used in tests return null if "off-row" - if this behaviour can be expected from all inputs
+        // there is no need to do anything here
+        private void pointToCombinedColumns() {
+        }
+
+        private void pointToNullColumns() {
+        }
+    }
+
+    /**
+     * Nested Loop + additional loop afterwards to emit any rows that had no matches
+     *
+     * <pre>
+     *     for (leftRow in left) {
+     *         for (rightRow in right) {
+     *             if matched {
+     *                 markPosition(pos)
+     *                 onRow
+     *             }
+     *         }
+     *     }
+     *
+     *     for (rightRow in right) {
+     *         if (noMatch(position)) {
+     *             onRow (left-side-null)
+     *         }
+     *     }
+     * </pre>
+     */
+    private static class RightJoinBatchIterator extends NestedLoopBatchIterator {
+
+        private final BitSet matchedRows = new BitSet();
```

Comment:
We should probably add lucene as a dependency and use the LongBitSet, otherwise we're limited by Integer.MAX_VALUE rows on the right side?